sql/binlog_crypt_data.cc

```diff
@@ -0,0 +1,126 @@
+#include "binlog_crypt_data.h"
+
+#include "my_global.h"
+#include "my_sys.h"
+#ifdef MYSQL_SERVER
+#include <mysql/service_mysql_keyring.h>
+#endif
+#include <algorithm>
+#include "boost/move/unique_ptr.hpp"
+
+Binlog_crypt_data::Binlog_crypt_data()
+  : key(NULL)
+  , enabled(false)
+  , scheme(0)
+{}
+
+Binlog_crypt_data::~Binlog_crypt_data()
+{
+  free_key();
+}
+
+Binlog_crypt_data::Binlog_crypt_data(const Binlog_crypt_data &b)
+{
+  enabled = b.enabled;
+  key_version = b.key_version;
+  if (b.key_length && b.key != NULL)
+  {
+    key= reinterpret_cast<uchar*>(my_malloc(PSI_NOT_INSTRUMENTED, b.key_length, MYF(MY_WME)));
+    memcpy(key, b.key, b.key_length);
+  }
+  else
+    key= NULL;
+
+  key_length= b.key_length;
+  memcpy(iv, b.iv, BINLOG_IV_LENGTH);
+  dst_len = b.dst_len;
+  memcpy(nonce, b.nonce, BINLOG_NONCE_LENGTH);
+}
+
+void Binlog_crypt_data::free_key()
+{
+  if (key != NULL)
+  {
+    DBUG_ASSERT(key_length == 16);
+    memset_s(key, 512, 0, key_length);
+    my_free(key);
+    key= NULL;
+    key_length= 0;
+  }
+}
+
+Binlog_crypt_data& Binlog_crypt_data::operator=(Binlog_crypt_data b)
+{
+  this->swap(b);
+  return *this;
+}
+
+void Binlog_crypt_data::swap(Binlog_crypt_data &b)
```

Comment:
These are not swap semantics - confusing, but I am not sure whether it's worth fixing it (and proper fix would probably be easiest by going back to assignment operator implementation)

Reply:
Yes, I agree. Moving back this function just to assign operator would be best. Having implement this swap function in proper way requires real swaps that are not needed in our case. I will put it back into assignment operator.